sky/data/data_utils.py

```diff
@@ -194,6 +194,58 @@ def parallel_upload(source_path_list: List[str],
                 [bucket_name] * len(commands)))
 
 
+def add_gsutil_platform_flags(gsutil_cmd: str) -> str:
+    """Adds platform-specific flags for gsutil.
+
+    In particular, we disable multiprocessing on Mac using
+    `-o "GSUtil:parallel_process_count=1"`. Multithreading is still enabled.
+    gsutil on Mac has a bug with multiprocessing that causes it to crash
+    when uploading files. Related issues:
+    https://bugs.python.org/issue33725
+    https://github.com/GoogleCloudPlatform/gsutil/issues/464
+
+    The flags are added by checking the platform using bash in a one-liner.
+    The platform check is done inline to have the flags match where the command
+    is executed, rather than where the code is run. This is important when
+    the command is run in a remote VM.
+
+    Args:
+        gsutil_cmd (str): The gsutil command to which platform-specific
+          flags will be added.
+
+    Returns:
+        str: Platform-specific command for running gsutil_cmd. Includes an if
+          conditional, and is run with `eval` before the command to
+          correctly run the if conditional.
+    """
+    disable_multiprocessing_flag = '-o "GSUtil:parallel_process_count=1"'
+
+    # Split the input command into parts
+    cmd_parts = gsutil_cmd.split()
+
+    # Check if the '-m' flag is in the command
+    if '-m' in cmd_parts:
+        # Find the index of '-m' flag and
+        # insert the platform-specific flag after it
+        index = cmd_parts.index('-m')
+        cmd_parts.insert(index + 1, disable_multiprocessing_flag)
```

Comment:
Ah yeah, this is what I tried first.. unfortunately it doesn't work because:
* Adding a flag with inline command substituion requires running the command with `eval` - see this [example output](https://gist.github.com/romilbhardwaj/c0823c6d353c2910a65b3afc0fefda93).
* However, if we use `eval`,  it also tries to evaluate with the `-x 'git/*'` flag we use for excluding git directories - [example output](https://gist.github.com/romilbhardwaj/c0823c6d353c2910a65b3afc0fefda93?permalink_comment_id=4614652#gistcomment-4614652).
* I tried escaping the -x flags, but couldn't figure out how to :(

To keep it simple, I decided to drop inline substitution in the command and instead use a if condition. Any ideas how to work around this?

Reply:
Ahh, sorry, I missed the comment above. That makes sense. I wondering if we can do something like following with the alias:

```
def get_gsutil_comand_with_arguments() -> Tuple[str, str]:
    cmd = f'$([[ "$(uname)" == "Darwin" ]] && alias skypilot_gsutil="gsutil -m {disable_multiprocessing_flag!r}" || alias skypilot_gsutil="gsutil -m")'
    return 'skypilot_gsutil', cmd

gsutil_executable, cmd = get_gsutil_command_with_arguments()
gsutil_cmd = f'{cmd}; myfile.txt | {gsutil_executable} cp -i xx xx'
```

This might be more structured than doing the postprocessing for the command, but I will let you decide. : )

Reply:
Unfortunately this also does not work, since the alias gets defined only within the subshell environment created by the command substitution. Once the command substitution is complete, the alias is no longer available in the main shell :(

```
(base) ➜  ~ $(alias skypilot_gsutil="gsutil -m")
(base) ➜  ~ skypilot_gsutil
zsh: command not found: skypilot_gsutil
```

Happy to try any other ideas, else we can perhaps go ahead with slicing the command and inserting flags manually.